sys/net/network_layer/ng_sixlowpan/frag/ng_sixlowpan_frag.c

```diff
@@ -0,0 +1,233 @@
+                                     size_t size)
+{
+    ng_netif_hdr_t *hdr = pkt->data, *new_hdr;
+    ng_pktsnip_t *netif, *frag;
+
+    netif = ng_netif_hdr_build(ng_netif_hdr_get_src_addr(hdr), hdr->src_l2addr_len,
+                               ng_netif_hdr_get_dst_addr(hdr), hdr->dst_l2addr_len);
+
+    if (netif == NULL) {
+        DEBUG("6lo frag: error allocating new link-layer header\n");
+        return NULL;
+    }
+
+    new_hdr = netif->data;
+    new_hdr->if_pid = hdr->if_pid;
+    new_hdr->flags = hdr->flags;
+    new_hdr->rssi = hdr->rssi;
+    new_hdr->lqi = hdr->lqi;
+
+    frag = ng_pktbuf_add(NULL, NULL, _min(size, payload_len),
+                         NG_NETTYPE_SIXLOWPAN);
+
+    if (frag == NULL) {
+        DEBUG("6lo frag: error allocating first fragment\n");
+        ng_pktbuf_release(netif);
+        return NULL;
+    }
+
+    LL_PREPEND(frag, netif);
+
+    return frag;
+}
+
+static uint16_t _send_1st_fragment(ng_sixlowpan_netif_t *iface, ng_pktsnip_t *pkt,
+                                   size_t payload_len, size_t datagram_size)
+{
+    ng_pktsnip_t *frag;
+    uint16_t max_frag_size = _floor8(iface->max_frag_size -
+                                     (payload_len - datagram_size) -
+                                     sizeof(ng_sixlowpan_frag_t));
+    uint16_t local_offset = 0;
+    ng_sixlowpan_frag_t *hdr;
+    uint8_t *data;
+
+    DEBUG("6lo frag: determined max_frag_size = %" PRIu16 "\n", max_frag_size);
+
+    /* 6LoWPAN dispatches don't count into that */
+    max_frag_size += (payload_len - datagram_size);
+
+    frag = _build_frag_pkt(pkt, payload_len,
+                           max_frag_size + sizeof(ng_sixlowpan_frag_t));
+
+    if (frag == NULL) {
+        return 0;
+    }
+
+    hdr = frag->next->data;
+    data = (uint8_t *)(hdr + 1);
+
+    hdr->disp_size = byteorder_htons((uint16_t)datagram_size);
+    hdr->disp_size.u8[0] |= NG_SIXLOWPAN_FRAG_1_DISP;
+    hdr->tag = byteorder_htons(_tag);
+
+    pkt = pkt->next;    /* don't copy netif header */
+
+    while (pkt != NULL) {
+        size_t clen = _min(max_frag_size - local_offset, pkt->size);
+
+        memcpy(data + local_offset, pkt->data, clen);
+        local_offset += clen;
+
+        if (local_offset >= max_frag_size) {
+            break;
+        }
+
+        pkt = pkt->next;
+    }
+
+    DEBUG("6lo frag: send first fragment (datagram size: %u, "
+          "datagram tag: %" PRIu16 ", fragment size: %" PRIu16 ")\n",
+          (unsigned int)datagram_size, _tag, local_offset);
+    ng_netapi_send(iface->pid, frag);
+
+    return local_offset;
+}
+
+static uint16_t _send_nth_fragment(ng_sixlowpan_netif_t *iface, ng_pktsnip_t *pkt,
+                                   size_t payload_len, size_t datagram_size,
+                                   uint16_t offset)
+{
+    ng_pktsnip_t *frag;
+    uint16_t max_frag_size = _floor8(iface->max_frag_size - sizeof(ng_sixlowpan_frag_n_t));
```

Comment:
because the difference of `payload_len` and `datagram_size` is encoded in the offset now.